### PR TITLE
[core] use the common `RouterIdSet` class

### PR DESCRIPTION
**Pull request by librasungirl**
Currently there are four similar copies regarding Router Id set separately in 1) `RouteTlv` in mle_tlvs.hpp; 2) `ThreadRouterMaskTlv` in thread_tlvs.hpp; 3) `RouterIdSet` in router_table.hpp; 4)`RouteTlv` in network_diagnostic_tlvs.hpp.

This PR extracts the common `RouterIdSet` class and apply it in the four components, meanwhile keep one copy for `RouteTlv` and `ThreadRouterMaskTlv` process in route_table (see [comments in #4749](https://github.com/openthread/openthread/pull/4749#discussion_r399736521))

### CODE REVIEW

File: src/core/thread/mle_types.hpp [hunk 1]
```diff
@@ -567,6 +568,61 @@ class LeaderData : public otLeaderData
...
+    /**
+     * This method removes the Router ID.
+     *
+     * @param[in]  aRouterId  The Router ID to remove.
+     *
+     */
+    void Remove(uint8_t aRouterId) { mRouterIdSet[aRouterId / 8] &= ~(0x80 >> (aRouterId % 8)); }
+
+    /**
+     * This method returns whether or not two Router ID sets equal.
```
Comment (jwhui) [inline src/core/thread/mle_types.hpp]:
```suggestion
     * This method returns whether or not two Router ID sets are equal.
```

File: src/core/thread/mle_types.hpp [hunk 2]
```diff
@@ -567,6 +568,61 @@ class LeaderData : public otLeaderData
...
+     *
+     * @param[in]  aOther The other Router ID Set to compare with.
+     *
+     * @retval TRUE   If two Router ID sets equal.
```
Comment (jwhui) [inline src/core/thread/mle_types.hpp]:
```suggestion
     * @retval TRUE   If the two Router ID sets are equal.
```

File: src/core/thread/mle_types.hpp [hunk 3]
```diff
@@ -567,6 +568,61 @@ class LeaderData : public otLeaderData
...
+     * @retval FALSE  If two Router ID sets don't equal.
```
Comment (jwhui) [inline src/core/thread/mle_types.hpp]:
```suggestion
     * @retval FALSE  If the Router ID sets are not equal.
```

File: src/core/thread/mle_types.hpp [hunk 4]
```diff
@@ -567,6 +568,61 @@ class LeaderData : public otLeaderData
...
+     * @param[in]  aRouterId  The Router ID.
+     *
+     * @retval TRUE   If the Router ID bit is set.
+     * @retval FALSE  If the Router ID bit is not set.
+     *
+     */
+    bool Contains(uint8_t aRouterId) const { return (mRouterIdSet[aRouterId / 8] & (0x80 >> (aRouterId % 8))) != 0; }
+
+    /**
+     * This method sets the Router ID.
```
Comment (jwhui) [inline src/core/thread/mle_types.hpp]:
```suggestion
     * This method sets a given Router ID.
```

File: src/core/thread/mle_types.hpp [hunk 5]
```diff
@@ -567,6 +568,61 @@ class LeaderData : public otLeaderData
...
+     *
+     * @param[in]  aRouterId  The Router ID to set.
+     *
+     */
+    void Add(uint8_t aRouterId) { mRouterIdSet[aRouterId / 8] |= 0x80 >> (aRouterId % 8); }
+
+    /**
+     * This method removes the Router ID.
```
Comment (jwhui) [inline src/core/thread/mle_types.hpp]:
```suggestion
     * This method removes a given Router ID.
```

File: src/core/thread/mle_types.hpp [hunk 6]
```diff
@@ -567,6 +568,61 @@ class LeaderData : public otLeaderData
...
+     *
+     * @param[in]  aOther The other Router ID Set to compare with.
+     *
+     * @retval TRUE   If two Router ID sets equal.
+     * @retval FALSE  If two Router ID sets don't equal.
+     *
+     */
+    bool operator!=(const RouterIdSet &aOther) const
+    {
+        return memcmp(mRouterIdSet, aOther.mRouterIdSet, OT_ARRAY_LENGTH(mRouterIdSet)) != 0;
```
Comment (jwhui) [inline src/core/thread/mle_types.hpp]:
```suggestion
        return memcmp(mRouterIdSet, aOther.mRouterIdSet, sizeof(mRouterIdSet)) != 0;
```
I think better to use `sizeof()` since we care more about the size rather than the number of elements.

File: src/core/thread/mle_types.hpp [hunk 7]
```diff
@@ -567,6 +568,61 @@ class LeaderData : public otLeaderData
...
 
+OT_TOOL_PACKED_BEGIN
+class RouterIdSet
+{
+public:
+    /**
+     * This method clears the Router Id Set.
+     *
+     */
+    void Clear(void) { memset(mRouterIdSet, 0, OT_ARRAY_LENGTH(mRouterIdSet)); }
```
Comment (jwhui) [inline src/core/thread/mle_types.hpp]:
```suggestion
    void Clear(void) { memset(mRouterIdSet, 0, sizeof(mRouterIdSet)); }
```
I think better to use `sizeof()` since we care more about the size rather than the number of elements.

File: src/core/thread/router_table.cpp [hunk 8]
```diff
@@ -306,7 +306,7 @@ otError RouterTable::Release(uint8_t aRouterId)
     return error;
 }
 
-void RouterTable::RemoveNeighbor(Router &aRouter)
```
Comment (simonlingoogle) [inline src/core/thread/router_table.cpp]: Is this change related to this PR?

Comment (librasungirl) [reply src/core/thread/router_table.cpp]: It might be not so related to the intention of this PR, however `RemoveNeighbor()` name does not exactly reflect what it does when updating `ProcessTlv()` to `ProcessRouterIdSet()`.
It is just a small piece of code, prefer to keep it as is.

Comment (simonlingoogle) [reply src/core/thread/router_table.cpp]: Agree that `RouterTable::RemoveNeighbor` does not reflect what it actually do. But, neither do I think `RemoveRouter` reflects exactly. 
`RemoveRouter` is basically same as `RemoveNeighbor`, because every `Neighbor` in `RouterTable` is `Router`. However, what this function actually removes is the **ROUTES** that use the router as the next hop. 

So, maybe this function should be renamed to `RemoveRoutes` or `RemoveRoutesFrom`? Thoughts? @jwhui @bukepo

Comment (jwhui) [reply src/core/thread/router_table.cpp]: FWIW, `RemoveNeighbor()` was closer than `RemoveRouter()`. This method removes direct link to the neighboring device (effectively removing the neighbor). Of course, as a result, this also removes any routes that utilize the direct neighbor link.

Comment (librasungirl) [reply src/core/thread/router_table.cpp]: The removed entry is what was in the router table but now not when learning the latest `RouterIdSet` either from `RouteTlv` or `ThreadRouterMaskTlv`, it may be the device's direct neighbor, but may also be some other routers which were routable.

As @jwhui , when removing one router entry, the other entries which was routed via this entry should be updated along with.

Thoughts?

Comment (jwhui) [reply src/core/thread/router_table.cpp]: This method is also in the call chain initiated by `MeshForwarder()` to invalidate a link to a neighboring device.
https://github.com/openthread/openthread/blob/d85c0007a13b05ceac5f2b070417d9f191bc5cdc/src/core/thread/mesh_forwarder.cpp#L813

If we want to rename, we could adopt the same name `RemoveRouterLink()`.

Comment (librasungirl) [reply src/core/thread/router_table.cpp]: Thought the neighboring router also belongs router, not conflicting with `RemoveRouter()` name.
Not sure but the `Link` in `RemoveRouterLink()` gives me the impression is it is removing direct neighbor, is it wrong?

Comment (jwhui) [reply src/core/thread/router_table.cpp]: This method only sets the link quality to zero and invalidates any routes that utilize the link. It does not actually remove any other neighbor or router state. `RemoveRouter()` might give the impression that the frame counters and other state are being removed. Thoughts?

Comment (librasungirl) [reply src/core/thread/router_table.cpp]: ok. that makes sense. will update.

File: src/core/thread/mle_types.hpp [hunk 9]
```diff
@@ -567,6 +568,61 @@ class LeaderData : public otLeaderData
     void SetLeaderRouterId(uint8_t aRouterId) { mLeaderRouterId = aRouterId; }
 };
 
+OT_TOOL_PACKED_BEGIN
+class RouterIdSet
```
Comment (jwhui) [inline src/core/thread/mle_types.hpp]: Would a more natural location for this be in `router_table.hpp`?

File: src/core/thread/mle_types.hpp [hunk 10]
```diff
@@ -567,6 +568,61 @@ class LeaderData : public otLeaderData
...
+     * @param[in]  aOther The other Router ID Set to compare with.
+     *
+     * @retval TRUE   If the Router ID sets are equal.
+     * @retval FALSE  If the Router ID sets are not equal.
+     *
+     */
+    bool operator!=(const RouterIdSet &aOther) const
+    {
+        return memcmp(mRouterIdSet, aOther.mRouterIdSet, sizeof(mRouterIdSet)) != 0;
+    }
```
Comment (abtink) [inline src/core/thread/mle_types.hpp]: How about if we also add `operator==` as well and implement `!=` from it instead?

File: src/core/thread/mle_types.hpp [hunk 11]
```diff
@@ -567,6 +568,61 @@ class LeaderData : public otLeaderData
...
+     *
+     */
+    void Remove(uint8_t aRouterId) { mRouterIdSet[aRouterId / 8] &= ~(0x80 >> (aRouterId % 8)); }
+
+    /**
+     * This method returns whether or not the Router ID sets are equal.
+     *
+     * @param[in]  aOther The other Router ID Set to compare with.
+     *
+     * @retval TRUE   If the Router ID sets are equal.
```
Comment (abtink) [inline src/core/thread/mle_types.hpp]: nit: update doc, since this is `!=` and returns `TRUE` if not equal.

File: src/core/thread/mle_types.hpp (hunk 9 above)
Comment (abtink) [reply src/core/thread/mle_types.hpp]: I also prefer the definition in `router_table.hpp`, but I think it may cause include dependency loops. If not easy to move it, I'd say ok to keep it in `mle_type.hpp`.

File: src/core/thread/router_table.cpp [hunk 12]
```diff
@@ -502,23 +502,25 @@ uint8_t RouterTable::GetLinkCost(Router &aRouter)
...
+
     for (uint8_t routerId = 0; routerId <= Mle::kMaxRouterId; routerId++)
     {
-        if (aTlv.IsRouterIdSet(routerId) == IsAllocated(routerId))
+        if (aRouterIdSet.Contains(routerId) == IsAllocated(routerId))
         {
             continue;
         }
 
         allocationChanged = true;
```
Comment (abtink) [inline src/core/thread/router_table.cpp]: Can we consider removing the `allocationChanged` variable (and the related `if` below), since we know after the newly added `VerifyOrExit(mAllocatedRouterIds != aRouterIdSet)` check, some router id has definitely changed.

File: src/core/thread/router_table.cpp [hunk 13]
```diff
@@ -502,23 +502,25 @@ uint8_t RouterTable::GetLinkCost(Router &aRouter)
...
-    mRouterIdSequence            = aTlv.GetRouterIdSequence();
+    mRouterIdSequence            = aRouterIdSequence;
     mRouterIdSequenceLastUpdated = TimerMilli::GetNow();
 
+    VerifyOrExit(mAllocatedRouterIds != aRouterIdSet);
+
     for (uint8_t routerId = 0; routerId <= Mle::kMaxRouterId; routerId++)
     {
-        if (aTlv.IsRouterIdSet(routerId) == IsAllocated(routerId))
+        if (aRouterIdSet.Contains(routerId) == IsAllocated(routerId))
```
Comment (abtink) [inline src/core/thread/router_table.cpp]: Can we remove this check as well? I think it does not serve any purpose if we remove `allocationChanged`?

File: src/core/thread/router_table.cpp [hunk 14]
```diff
@@ -502,23 +502,25 @@ uint8_t RouterTable::GetLinkCost(Router &aRouter)
...
         {
             continue;
         }
 
         allocationChanged = true;
 
-        if (aTlv.IsRouterIdSet(routerId))
+        if (aRouterIdSet.Contains(routerId))
         {
             mAllocatedRouterIds.Add(routerId);
```
Comment (abtink) [inline src/core/thread/router_table.cpp]: How about if remove the `mAllocatedRouterIds.Add(routerId)` and the `mAllocatedRouterIds.Remove(routerId);` in the `else`?  And instead after the loop, do a `mAllocatedRouterIds = aRouterIdSet` to copy all the bits at the same time.

This should hep make the the loop simple: We go through all the router id values and check the ones that are removed (are contained in the old `mAllocatedRouterIds` but not in the new `aRouterIdSet`) and do the `SetNextHop` and `RemoveRouterLink` for those...

File: src/core/thread/router_table.cpp [hunk 15]
```diff
@@ -539,39 +541,9 @@ void RouterTable::ProcessTlv(const Mle::RouteTlv &aTlv)
...
-
-        allocationChanged = true;
-
-        if (aTlv.IsAssignedRouterIdSet(routerId))
-        {
-            mAllocatedRouterIds.Add(routerId);
-        }
-        else
-        {
-            mAllocatedRouterIds.Remove(routerId);
```
Comment (abtink) [inline src/core/thread/router_table.cpp]: There is a difference in behavior here where we don't do 
```
     {
            Router *router = GetRouter(routerId);

            OT_ASSERT(router != NULL);
            router->SetNextHop(Mle::kInvalidRouterId);
            RemoveRouterLink(*router);
     }
```
compared to the new common `ProcessRouterIdSet()`. 

I am not sure if it matters. I think it's fine to use common method. @jwhui , @librasungirl  do you agree?

File: src/core/thread/router_table.hpp [hunk 16]
```diff
@@ -312,20 +312,21 @@ class RouterTable : public InstanceLocator
...
-     * This method updates the router table with a received Route TLV.
+     * This method updates the Router ID allocation.
      *
-     * @param[in]  aTlv  A reference to the Route TLV.
+     * @param[in]  aRouterIdSequence  The Router Id Sequence.
+     * @param[in]  aRouterIdSet       A reference to the Router Id Set.
      *
      */
-    void ProcessTlv(const Mle::RouteTlv &aTlv);
+    void ProcessRouterIdSet(uint8_t aRouterIdSequence, const Mle::RouterIdSet &aRouterIdSet);
```
Comment (abtink) [inline src/core/thread/router_table.hpp]: How about renaming this to `UpdateRouterIdSet()`?
(it is fine for me as is, just a suggestion).

File: src/core/thread/router_table.cpp (hunk 15 above)
Comment (jwhui) [reply src/core/thread/router_table.cpp:566]: Good point. I think the added logic in this case is OK.

File: src/core/thread/mle_types.hpp (hunk 9 above)
Comment (librasungirl) [reply src/core/thread/mle_types.hpp:572]: yes, like @abtink mentioned, putting it here was to avoid include dependency loops.

File: src/core/thread/router_table.cpp (hunk 12 above)
Comment (librasungirl) [reply src/core/thread/router_table.cpp:522]: yes, you are right!  updated.

File: src/core/thread/router_table.hpp (hunk 16 above)
Comment (librasungirl) [reply src/core/thread/router_table.hpp:321]: renamed as your suggestion.

File: src/core/thread/router_table.cpp (hunk 14 above)
Comment (librasungirl) [reply src/core/thread/router_table.cpp:525]: good point!

File: src/core/thread/mle_types.hpp (hunk 10 above)
Comment (librasungirl) [reply src/core/thread/mle_types.hpp:620]: updated as your suggestion.
